controller/volume_attachment_controller.go

```diff
@@ -283,129 +292,192 @@ func (vac *VolumeAttachmentController) handleVolumeAttachment(va *longhorn.Volum
 	}
 
-	// Attachments that desires detaching
-	for _, attachment := range va.Status.Attachments {
-		if _, ok := va.Spec.Attachments[attachment.ID]; !ok {
-			updateStatusForDesiredDetachingAttachment(attachment, va, vol)
+	// Attachment that desires detaching
+	for _, attachmentStatus := range va.Status.AttachmentStatuses {
+		if _, ok := va.Spec.AttachmentSpecs[attachmentStatus.ID]; !ok {
+			updateStatusForDesiredDetachingAttachment(attachmentStatus.ID, va)
 		}
 	}
 
-	// Attachments that are requesting to attach
-	for _, attachment := range va.Spec.Attachments {
-		updateStatusForDesiredAttachingAttachment(attachment, va, vol)
+	// Attachment that requests to attach
+	for _, attachmentSpec := range va.Spec.AttachmentSpecs {
+		updateStatusForDesiredAttachingAttachment(attachmentSpec.ID, va, vol)
 	}
 	return nil
 }
 
-func updateStatusForDesiredDetachingAttachment(attachment *longhorn.Attachment, va *longhorn.VolumeAttachment, vol *longhorn.Volume) {
+func updateStatusForDesiredDetachingAttachment(attachmentID string, va *longhorn.VolumeAttachment) {
 	//TODO: How to handle vol.Status.IsStandby volume
-	delete(va.Status.Attachments, attachment.ID)
+	delete(va.Status.AttachmentStatuses, attachmentID)
 }
 
-func updateStatusForDesiredAttachingAttachment(specAttachment *longhorn.Attachment, va *longhorn.VolumeAttachment, vol *longhorn.Volume) {
-	if _, ok := va.Status.Attachments[specAttachment.ID]; !ok {
-		va.Status.Attachments[specAttachment.ID] = &longhorn.Attachment{
-			ID:         specAttachment.ID,
-			Type:       specAttachment.Type,
-			NodeID:     specAttachment.NodeID,
-			Parameters: copyStringMap(specAttachment.Parameters),
-			Attached:   utilpointer.Bool(false),
+func updateStatusForDesiredAttachingAttachment(attachmentID string, va *longhorn.VolumeAttachment, vol *longhorn.Volume) error {
+	if _, ok := va.Status.AttachmentStatuses[attachmentID]; !ok {
+		va.Status.AttachmentStatuses[attachmentID] = &longhorn.AttachmentStatus{
+			ID: attachmentID,
+			// TODO: handle condition initialization here
 		}
 	}
-	statusAttachment := va.Status.Attachments[specAttachment.ID]
 
-	// Sync info from the corresponding attachment in va.Spec.Attachments if there are changes
-	if statusAttachment.Type != specAttachment.Type ||
-		statusAttachment.NodeID != specAttachment.NodeID ||
-		!reflect.DeepEqual(statusAttachment.Parameters, specAttachment.Parameters) {
-		// sync info from spec
-		statusAttachment.Type = specAttachment.Type
-		statusAttachment.NodeID = specAttachment.NodeID
-		statusAttachment.Parameters = copyStringMap(specAttachment.Parameters)
-		// reset status info
-		statusAttachment.Attached = utilpointer.Bool(false)
-		statusAttachment.AttachError = nil
-		statusAttachment.DetachError = nil
+	attachmentSpec, ok := va.Spec.AttachmentSpecs[attachmentID]
+	if !ok {
+		return fmt.Errorf("updateStatusForDesiredAttachingAttachment: missing the attachment with id %v in the va.Spec.AttachmentSpecs", attachmentID)
```

Comment:
Need to print out this somewhere. The caller does not handle errors of this function.

Reply:
Fixed it, Thank you